README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# dmhct

<!-- badges: start -->

[![R-CMD-check](https://github.com/jesse-smith/dmhct/actions/workflows/R-CMD-check.yaml/badge.svg)](https://github.com/jesse-smith/dmhct/actions/workflows/R-CMD-check.yaml)
<!-- badges: end -->

`{dmhct}` extracts, loads, and transforms data for the MLinHCT project.
The main functionality is wrapped into a single function, `dm_elt()`,
that produces a `dm` object (from the [`{dm}`
package](https://cynkra.github.io/dm/)) containing cleaned data tables
from an SQL Server database.

## Installation and Setup

You can install `{dmhct}` from Github with:

``` r
# install.packages("remotes")
remotes::install_github("jesse-smith/dmhct")
```

To use the package, you will need access to the SQL Server instance
containing the data. This is provided by St. Jude IT; contact [Jesse
Smith](mailto:jesse.smith@stjude.org) to learn how to gain access.

## Introduction

The main functionality in the package is in `dm_elt()`; this function
combines two sub-functions, `dm_extract()` and `dm_transform()`, into a
small pipeline.

``` r
library(dmhct)
dm <- dm_elt()
#> Upgrading dm object created with dm <= 0.3.0.
#> Upgrading dm object created with dm <= 0.3.0.
dm2 <- dm_extract() %>% dm_transform()
#> Upgrading dm object created with dm <= 0.3.0.
#> Upgrading dm object created with dm <= 0.3.0.
identical(dm, dm2)
#> [1] TRUE
```

The `dm_extract()` function is itself a wrapper for a pipeline used to
extract the necessary data from the SQL Server and load it onto your
local machine. The `dm_transform()` function provides a second pipeline
that cleans each table in the database and outputs tables ready for
analysis (as [`data.table`s](https://rdatatable.gitlab.io/data.table/)).

Both functions cache their results by default and compare a checksum of
their inputs to a cached checksum to determine whether re-calculation is
necessary. This allows the functions to simply read the cached outputs,
rather than repeat costly computations, if the input is unchanged. Each
function also allows the user to force re-computation and cache
overwriting via a `reset` argument. This caching behavior (and the
control arguments) are made available in the `dm_elt()` wrapper as well.

## Tables

The data used in the project consists of 10 tables:

``` r
summary(dm)[, -3L]
#> Upgrading dm object created with dm <= 0.3.0.
#> Upgrading dm object created with dm <= 0.3.0.
#> Upgrading dm object created with dm <= 0.3.0.
#> Upgrading dm object created with dm <= 0.3.0.
#> Upgrading dm object created with dm <= 0.3.0.
#> Upgrading dm object created with dm <= 0.3.0.
#> Upgrading dm object created with dm <= 0.3.0.
#> Upgrading dm object created with dm <= 0.3.0.
#> Upgrading dm object created with dm <= 0.3.0.
#> Upgrading dm object created with dm <= 0.3.0.
#>                Length Class     
#> hla            14     data.table
#> master         26     data.table
#> cerner          5     data.table
#> chimerism       5     data.table
#> death          10     data.table
#> disease_status  4     data.table
#> engraftment     5     data.table
#> gvhd            5     data.table
#> mrd             5     data.table
#> relapse         4     data.table
```

The entry point to this data is the `master` table, which contains one
row per patient and includes baseline variables such as transplant date,
patient demographics, and donor characteristics. It also contains unique
IDs for patients (`entity_id`) and donors (`donor_id`).

``` r
skimr::skim(dm$master, setdiff(colnames(dm$master), c("entity_id", "donor_id")))
#> Upgrading dm object created with dm <= 0.3.0.
#> Upgrading dm object created with dm <= 0.3.0.
#> Upgrading dm object created with dm <= 0.3.0.
#> Upgrading dm object created with dm <= 0.3.0.
```

|                                                  |            |
|:-------------------------------------------------|:-----------|
| Name                                             | dm\$master |
| Number of rows                                   | 852        |
| Number of columns                                | 26         |
| Key                                              | entity_id  |
| \_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_   |            |
| Column type frequency:                           |            |
| character                                        | 1          |
| Date                                             | 1          |
| factor                                           | 13         |
| logical                                          | 3          |
| numeric                                          | 6          |
| \_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_\_ |            |
| Group variables                                  | None       |

Data summary

**Variable type: character**

| skim_variable      | n_missing | complete_rate | min | max | empty | n_unique | whitespace |
|:-------------------|----------:|--------------:|----:|----:|------:|---------:|-----------:|
| cat_donor_relation |       285 |          0.67 |   4 |  11 |     0 |        7 |          0 |

**Variable type: Date**

| skim_variable | n_missing | complete_rate | min        | max        | median     | n_unique |
|:--------------|----------:|--------------:|:-----------|:-----------|:-----------|---------:|
| dt_trans      |         0 |             1 | 2000-01-15 | 2022-09-06 | 2012-02-02 |      807 |

**Variable type: factor**

| skim_variable               | n_missing | complete_rate | ordered | n_unique | top_counts                            |
|:----------------------------|----------:|--------------:|:--------|---------:|:--------------------------------------|
| cat_degree_match06          |        62 |          0.93 | TRUE    |        4 | 6: 499, 3: 138, 4: 83, 5: 70          |
| cat_degree_match08          |       230 |          0.73 | TRUE    |        5 | 8: 340, 4: 116, 5: 85, 7: 43          |
| cat_degree_match10          |       234 |          0.73 | TRUE    |        6 | 10: 320, 6: 86, 5: 82, 7: 53          |
| cat_disease_status_at_trans |       188 |          0.78 | FALSE   |        3 | Rem: 411, Act: 179, Rel: 74           |
| cat_donor_race              |       182 |          0.79 | FALSE   |        3 | Whi: 434, Bla: 146, Oth: 90           |
| cat_donor_sex               |        43 |          0.95 | FALSE   |        2 | Mal: 419, Fem: 390                    |
| cat_dx_grp                  |        10 |          0.99 | FALSE   |        9 | AML: 283, ALL: 267, Oth: 94, Apl: 58  |
| cat_ethnicity               |         8 |          0.99 | FALSE   |        2 | Not: 669, His: 175                    |
| cat_matched_related         |       368 |          0.57 | FALSE   |        4 | mis: 177, mat: 149, mat: 127, mis: 31 |
| cat_prep_type               |       224 |          0.74 | TRUE    |        3 | Mye: 373, Red: 219, Non: 36           |
| cat_product_type            |         4 |          1.00 | FALSE   |        2 | Mar: 481, PBS: 367, Cor: 0, Mar: 0    |
| cat_race                    |         1 |          1.00 | FALSE   |        3 | Whi: 597, Bla: 165, Oth: 89           |
| cat_sex                     |         0 |          1.00 | FALSE   |        2 | Mal: 487, Fem: 365                    |

**Variable type: logical**

| skim_variable     | n_missing | complete_rate | mean | count              |
|:------------------|----------:|--------------:|-----:|:-------------------|
| lgl_death         |         0 |          1.00 | 0.35 | FAL: 551, TRU: 301 |
| lgl_donor_related |       285 |          0.67 | 0.58 | TRU: 328, FAL: 239 |
| lgl_malignant     |        10 |          0.99 | 0.78 | TRU: 657, FAL: 185 |

**Variable type: numeric**

| skim_variable | n_missing | complete_rate |    mean |      sd |      p0 |     p25 |     p50 |     p75 |    p100 | hist  |
|:--------------|----------:|--------------:|--------:|--------:|--------:|--------:|--------:|--------:|--------:|:------|
| num_age       |         0 |          1.00 |    9.61 |    5.98 |    0.07 |    4.19 |    9.77 |   14.97 |   20.88 | ▇▆▆▇▅ |
| num_age_donor |        49 |          0.94 |   28.06 |   12.64 |    0.60 |   20.02 |   28.19 |   37.71 |   65.72 | ▃▇▇▅▁ |
| num_dt_trans  |         0 |          1.00 | 2011.78 |    6.65 | 2000.04 | 2005.71 | 2012.09 | 2017.87 | 2022.68 | ▆▇▆▆▇ |
| num_n\_trans  |         0 |          1.00 |    1.00 |    0.00 |    1.00 |    1.00 |    1.00 |    1.00 |    1.00 | ▁▁▇▁▁ |
| num_t\_surv   |         0 |          1.00 | 2342.64 | 2410.06 |  -48.00 |  271.75 | 1283.00 | 4044.00 | 8216.00 | ▇▂▂▂▁ |
| num_t\_trans  |        11 |          0.99 |  692.43 | 1060.01 |   16.00 |  123.00 |  232.00 |  748.00 | 6668.00 | ▇▁▁▁▁ |

All tables are keyed by one or more columns that are (jointly) unique to
each row. Each table contains `entity_id` as a primary key; most tables
except `master` then define additional keys specific to the data
contained in that table (for example, many tables contain repeated
observations for a patient and thus add a `date` key).

``` r
dm::dm_get_all_pks(dm) %>% dplyr::mutate(pk_col = format(.data$pk_col))
#> Upgrading dm object created with dm <= 0.3.0.
#> # A tibble: 10 × 2
#>    table          pk_col                      
#>    <chr>          <chr>                       
#>  1 hla            entity_id, donor_id         
#>  2 master         entity_id                   
#>  3 cerner         entity_id, test, date, n_obs
#>  4 chimerism      entity_id, date, cat_source 
#>  5 death          entity_id, n_id             
#>  6 disease_status entity_id, date, n_status   
#>  7 engraftment    entity_id, date             
#>  8 gvhd           entity_id, date, cat_site   
#>  9 mrd            entity_id, date             
#> 10 relapse        entity_id, date
```

Here’s a list of data (non-key) columns in each table (excluding
`master`):

``` r
tibble::tibble(
  table = names(dm::dm_select_tbl(dm, -"master")),
  cols = sapply(dm::dm_select_tbl(dm, -"master"), function(x) {
    paste0(setdiff(colnames(x), data.table::key(x)), collapse = ", ")
  })
)
#> Upgrading dm object created with dm <= 0.3.0.
#> Upgrading dm object created with dm <= 0.3.0.
#> Upgrading dm object created with dm <= 0.3.0.
#> Upgrading dm object created with dm <= 0.3.0.
#> Upgrading dm object created with dm <= 0.3.0.
#> Upgrading dm object created with dm <= 0.3.0.
#> # A tibble: 9 × 2
#>   table          cols                                                           
#>   <chr>          <chr>                                                          
#> 1 hla            n06, n08, n10, a, b, drb1, c, dqb1, dpb1, drb345, dqa1, dpa1   
#> 2 cerner         result                                                         
#> 3 chimerism      cat_method, pct_donor                                          
#> 4 death          dt_death, dt_trans, lgl_on_therapy, cat_death_location, cat_de…
#> 5 disease_status disease_status                                                 
#> 6 engraftment    recovery_anc500, recovery_plat20k, recovery_plat50k            
#> 7 gvhd           cat_grade, cat_type                                            
#> 8 mrd            cat_method, pct_result, lgl_result                             
#> 9 relapse        lgl_remission, dt_remission
```

## Development

You can find code generating results for each table in a corresponding
file named `"dm_[table-name]"`. Changes should not be committed directly
to the `main` branch; going forward, all changes should be made in a
feature/bug fix/etc. branch and reviewed by another team member before
merging. The Github repository is set up to dis-allow pushing direct
commits to `main`.

## Using with `{predHCT}`

`{dmhct}` was created to de-couple data cleaning code from modeling code
for the MLinHCT project. However, it is a required dependency for the
main modeling package, `{predHCT}`. While it is possible to install your
local development version for use with `{predHCT}`, this is not
recommended. Instead, please install from the `main` Github branch using
the installation instructions at the beginning of this document.


NEWS.md
<!-- NEWS.md is maintained by https://cynkra.github.io/fledge, do not edit -->

# dmhct 0.0.0.9002 (2022-11-02)

* Add `timestamp` attributes for each table and use for `num_t_surv` in


# dmhct 0.0.0.9001 (2022-11-02)

- Same as previous version.




R/dm_cerner.R
#' Unite All Cerner Tables and Remove Individual Tables
#'
#' @param dm_remote `[dm]` Remote `dm` connected to an SQL Server w/ HCT data
#' @param path_var `[chr(1)]` Path to CSV file mapping test names to
#'   standardized variable names
#' @param quiet `[lgl(1)]` Should the function return quietly if a Cerner table
#'   already exists?
#'
#' @return The `dm` w/ instructions for creating a combined `cerner` table
dm_cerner_extract <- function(
    dm_remote,
    path_var = system.file("extdata", "cerner_var_map.csv", package = "dmhct"),
    quiet = TRUE
) {
  if ("cerner" %in% names(dm_remote)) {
    if (!quiet) rlang::inform("`cerner` table already created")
    return(dm_remote)
  }
  # Select cerner tables
  dm_cerner <- dm::dm_select_tbl(dm_remote, dplyr::matches("(?i)cerner[0-9]"))
  # Create list of names for removal
  nm_list   <- as.list(names(dm_cerner))
  # Row bind to single table
  tbl_cerner <- purrr::reduce(
    dm_cerner[-1L],
    ~ dplyr::union_all(.x, utils_cerner$std_cerner_tbl(.y, dm_remote$master)),
    .init = utils_cerner$std_cerner_tbl(dm_cerner[[1L]], dm_remote$master)
  )
  # Add timestamp (max timestamp of individual tables)
  attr(tbl_cerner, "timestamp") <- dm_cerner %>%
    dm::dm_get_tables() %>%
    purrr::map(attr, "timestamp") %>%
    dplyr::as_tibble() %>%
    as.matrix() %>%
    max(na.rm = TRUE) %>%
    lubridate::as_datetime()

  dm_remote %>%
    # Add combined table
    dm::dm_add_tbl(cerner = tbl_cerner) %>%
    # Remove individual tables
    dm::dm_rm_tbl(!!!nm_list)
}


#' Transform an Extracted Cerner Table in a Local `dm`
#'
#' @param dm_local `[dm]` Local `dm` w/ HCT data
#' @param numeric_only `[lgl(1)]` Should non-numeric data be dropped during
#'   transformation?
#' @param path `[chr(1)]` Path to CSV file mapping test names to
#'   standardized variable names
#'
#' @return The `dm` object w/ transformed `cerner` table
#'
#' @export
dm_cerner_transform <- function(
    dm_local,
    numeric_only = TRUE,
    path = system.file("extdata", "cerner_var_map.csv", package = "dmhct")
) {
  dt <- dm_local$cerner
  class <- df_class(dt)
  dt <- data.table::as.data.table(dt)

  # Filter unused patients w/ master
  dt <- dt[entity_id %in% dm_local$master$entity_id]


  # Silence R CMD CHECK Notes
  entity_id <- var <- test <- result <- ..pos <- ..neg <- NULL

  # Load variable map
  var_map <- path %>%
    path_create() %>%
    data.table::fread(data.table = TRUE) %>%
    dplyr::distinct(.data$test, .keep_all = TRUE)

  # Standardize test names w/ variable map
  dt <- var_map[dt, on = "test"][, "test" := var][, "var" := NULL]

  # Remove non-numeric columns if numeric only is TRUE
  if (numeric_only) dt <- dt[test %flike% "num_" | test %flike% "pct_"]

  # Squish result strings
  dt[result %flike% " ", "result" := stringr::str_squish(result)]

  # Extract numbers from numeric variable results
  dt[
    test %flike% "num_" | test %flike% "pct_",
    "result" := data.table::fifelse(
      # Attempt numeric extraction for strings starting w/ number or "<>"
      result %like% "^[0-9.<>]",
      # Extract numbers + decimals
      stringr::str_extract(result, "([0-9]+)|([0-9]+?[.][0-9]+)"),
      NA_character_
    )
  ]

  # Threshold-based outlier removal
  dt[
    test == "num_ht",
    "result" := data.table::fifelse(as.numeric(result) %between% c(24, 272), result, NA_character_)
  ][
    test == "num_wt",
    "result" := data.table::fifelse(as.numeric(result) %between% c(1, 442), result, NA_character_)
  ][
    test == "num_bmi",
    "result" := data.table::fifelse(as.numeric(result) %between% c(10, 60), result, NA_character_)
  ][
    test == "num_temp",
    "result" := data.table::fifelse(as.numeric(result) <= 46, result, NA_character_)
  ][
    test == "num_sbp",
    "result" := data.table::fifelse(as.numeric(result) %between% c(75, 175), result, NA_character_)
  ][
    test == "num_dbp",
    "result" := data.table::fifelse(as.numeric(result) %between% c(30, 100), result, NA_character_)
  ][
    test == "num_rr",
    "result" := data.table::fifelse(as.numeric(result) %between% c(0, 75), result, NA_character_)
  ][
    test == "num_hr",
    "result" := data.table::fifelse(as.numeric(result) %between% c(0, 250), result, NA_character_)
  ][
    test == "pct_o2_sat",
    "result" := data.table::fifelse(as.numeric(result) %between% c(90, 100), result, NA_character_)
  ][
    test == "num_map",
    "result" := data.table::fifelse(as.numeric(result) <= 120, result, NA_character_)
  ][
    test == "num_mpv",
    "result" := data.table::fifelse(as.numeric(result) %between% c(4.5, 13.5), result, NA_character_)
  ][
    test == "num_mch",
    "result" := data.table::fifelse(as.numeric(result) <= 40, result, NA_character_)
  ][
    test == "num_rbc",
    "result" := data.table::fifelse(as.numeric(result) <= 5.5, result, NA_character_)
  ][
    test == "num_mchc",
    "result" := data.table::fifelse(as.numeric(result) %between% c(30, 40), result, NA_character_)
  ]

  # Adenovirus infection
  pos <- c("DETECTED", "Detected", "3.92", "4.07", "4.37", "4.71")
  neg <- c("NOT DET", "Not Detected", "<3.47")
  dt[
    test == "lgl_pcr_adenovirus",
    "result" := data.table::fcase(
      result %chin% ..pos, "TRUE",
      result %chin% ..neg, "FALSE"
    )
  ]
  rm(pos, neg)

  # Aspergillus infection
  neg <- c("Negative", "NEG")
  dt[test == "lgl_antigen_aspergillus",
     "result" := data.table::fcase(
       result == "Positive", "TRUE",
       result %chin% ..neg,   "FALSE"
     )]
  rm(neg)

  # Pertussis infection
  dt[test == "lgl_pcr_b_pertussis",
     "result" := data.table::fifelse(
       result == "Detected",
       "TRUE",
       "FALSE"
     )]

  # HCV antibody
  pos <- c("Reactive", "REACTIVE")
  neg <- c("NON REAC", "Non Reactive", "Non-reactive", "Negative")
  dt[test == "lgl_anti_hcv",
     "result" := data.table::fcase(
       result %chin% pos, "TRUE",
       result %chin% neg, "FALSE"
     )]
  rm(pos, neg)

  # Hep C
  neg <- c("Negative", "Nonreactive")
  dt[
    test == "lgl_anti_hcv",
    "result" := data.table::fcase(
      result == "Reactive", "TRUE",
      result %chin% ..neg, "FALSE"
    )
  ]

  # Flu A
  pos <- c("Flu A H3", "Flu A", "Flu A H1-2009")
  dt[test == "lgl_pcr_flu_a",
     "result" := data.table::fcase(
       result %chin% pos,         "TRUE",
       result == "Not Detected", "FALSE"
     )]
  rm(pos)

  # Parainfluenza
  dt[test == "lgl_pcr_flu_a",
     "result" := data.table::fifelse(
       result == "Not Detected",
       "FALSE",
       "TRUE"
     )]

  # Pregnant
  neg <- c("Negative", "NEGATIVE")
  dt[test == "lgl_pregnancy",
     "result" := data.table::fcase(
       result == "Positive", "TRUE",
       result %chin% neg,   "FALSE"
     )]
  rm(neg)

  # Rectal surveillance
  pos <- c("Positive", "Detected")
  neg <- c("Inhibited", "Inhibition", "Negative", "Not Detected")
  dt[test == "lgl_rectal_culture",
     "result" := data.table::fcase(
       result %chin% ..pos, "TRUE",
       result %chin% ..neg, "FALSE"
     )]
  rm(pos, neg)

  # Categorical urine variables
  u_vars <- paste0(
    "cat_u_",
    c("bili", "ketones", "blood", "prot", "no2", "leuko", "glucose", "urobili")
  )
  dt[
    test %chin% u_vars,
    "result" := data.table::fcase(
      result %chin% c("NEG", "NEGATIVE", "Negative"), "0",
      result %chin% c("1+", "POS 1+", "1.0"), "1+",
      result %chin% c("POSITIVE", "Positive", "TRACE", "Trace"), "1+",
      result %chin% c("2+", "POS 2+", "0.2", "2.0"), "2+",
      result %chin% c("3+", "POS 3+", "3"), "3+",
      result %chin% c("4+", "POS 4+", "4.0"), "4+"
    )
  ]

  # Boolean variables w/ no parsing needed
  bool_vars <- paste0(
    "lgl_pcr_",
    c("b_pertussis", "hmpv", "rhino_enterovirus", "flu_b")
  )
  dt[
    test %chin% bool_vars,
    "result" := data.table::fifelse(result == "Detected", "TRUE", "FALSE")
  ]

  # Set keys
  pk <- c("entity_id", "test", "date")

  # Set order
  data.table::setorderv(dt)

  # Remove duplicates before date conversion
  dt <- dt[!duplicated(dt)]

  # Convert datetime to date
  dt[, "date" := lubridate::as_date(date)]

  # Remove missings
  dt <- dt[!(is.na(test) | is.na(result))]

  # Give non-unique observations a unique ID
  dt[, "n_obs" := seq_len(.N), by = pk]

  # Add n_obs to keys
  pk <- c(pk, "n_obs")

  # Set key and reorder cols
  data.table::setkeyv(dt, cols = pk)
  data.table::setcolorder(dt)

  # Convert back to tibble if input was not data.table
  dt <- dt_cast(dt, to = class)

  # Ensure timestamp is retained
  attr(dt, "timestamp") <- attr(dm_local$cerner, "timestamp")

  dm <- dm_local %>%
    dm::dm_rm_tbl("cerner") %>%
    dm::dm_add_tbl(cerner = dt[]) %>%
    # Add primary key
    dm::dm_add_pk("cerner", !!data.table::key(dt), check = TRUE)
}

#' Utility Functions for Cerner Table ELT
#'
#' @description
#' Collection of utility functions for cerner data
#'
#' @aliases utils_cerner
#'
#' @keywords internal
UtilsCerner <- R6Class(
  "UtilsCerner",
  cloneable = FALSE,
  public = list(
    #' Standardize a Cerner Table
    #'
    #' @description
    #' Selects needed columns from a Cerner table and converts them to the
    #' expected class.
    #'
    #' @param tbl `[tbl_dbi]` A table containing Cerner data
    #' @param master `[tbl_dbi]` A table containing master `entity_id` data
    #'
    #' @return `[tbl_dbi]` The standardized table
    std_cerner_tbl = function(tbl, master = NULL) {
      # Get column names
      cols <- colnames(tbl)
      # Select and rename needed columns
      EntityId <- stringr::str_subset(cols, "(?i)entity")
      Date     <- stringr::str_subset(cols, "(?i)date")
      Test     <- stringr::str_subset(cols, "(?i)test")
      Result   <- stringr::str_subset(cols, "(?i)result")

      # Update tbl
      tbl %>%
        dplyr::select(
          entity_id = {{ EntityId }},
          date      = {{ Date }},
          test      = {{ Test }},
          result    = {{ Result }}
        ) %>%
        # Filter to patients in master if supplied
        dplyr::mutate(entity_id = as.integer(.data$entity_id)) %>%
        purrr::when(
          is.data.frame(master) ~ dplyr::semi_join(., master, by = "entity_id"),
          ~ .
        ) %>%
        # Ensure columns are of expected type
        dplyr::mutate(
          date = dbplyr::sql("CONVERT(DATETIME, date)"),
          test = trimws(as.character(.data$test)),
          result = trimws(as.character(.data$result))
        ) %>%
        # Filter
        dplyr::filter(
          !is.na(.data$entity_id),
          !is.na(.data$date),
          !is.na(.data$test),
          !is.na(.data$result)
        )
    }
  )
)

#' @rdname UtilsCerner
#' @usage NULL
#' @format NULL
#' @keywords internal
utils_cerner <- UtilsCerner$new()
